File: Advanced/Arrays/Fundamentals/6_left_rotate_arr_by_k_places.py
# Left Rotate Array by K places


# Brute Force Approach
# TC = O(n+k)
# SC = O(k)
def rotateArrayBrute(nums, k):
    n = len(nums)
    a = k % n
    temp = []
    for i in range(a):
        temp.append(nums[i])
    for i in range(a, len(nums)):
        nums[i - a] = nums[i]
    for i in range(a):
        nums[n - a + i] = temp[i]


# Optimal Approach
# TC = O(n)
# SC = O(1)
def rotateArrayOptimal(nums, k):
    n = len(nums)
    a = k % n
    reverse(nums, 0, a - 1)
    reverse(nums, a, n - 1)
    reverse(nums, 0, n - 1)


def reverse(arr, start, end):
    while start < end:
        arr[start], arr[end] = arr[end], arr[start]
        start += 1
        end -= 1


nums = [1, 2, 3, 4, 5]
k = 2
rotateArrayBrute(nums, k)
print(nums)

File: Advanced/Arrays/Fundamentals/6_left_rotate_arr_by_k_places.cpp
#include <bits/stdc++.h>
using namespace std;

// --- Brute Force Approach ---
// TC: O(n + k)
// SC: O(k)
void leftRotateBrute(vector<int> &nums, int k)
{
    int n = nums.size();
    int a = k % n;

    vector<int> temp;
    for (int i = 0; i < a; i++)
        temp.push_back(nums[i]);

    for (int i = a; i < n; i++)
        nums[i - a] = nums[i];

    for (int i = 0; i < a; i++)
        nums[n - a + i] = temp[i];
}

// --- Helper Function for Optimal Approach ---
void reverseVec(vector<int> &arr, int start, int end)
{
    while (start < end)
    {
        swap(arr[start], arr[end]);
        start++;
        end--;
    }
}

// --- Optimal Approach ---
// TC: O(n), SC: O(1)
void leftRotateOptimal(vector<int> &nums, int k)
{
    int n = nums.size();
    int a = k % n;

    reverseVec(nums, 0, a - 1);
    reverseVec(nums, a, n - 1);
    reverseVec(nums, 0, n - 1);
}

int main()
{
    vector<int> nums = {1, 2, 3, 4, 5};
    int k = 2;

    // leftRotateBrute(nums, k);
    leftRotateOptimal(nums, k);

    for (int num : nums)
    {
        cout << num << " ";
    }
    cout << endl;

    return 0;
}
